tsdb/store.go

```diff
@@ -1709,67 +1709,87 @@ func (s *Store) TagValues(ctx context.Context, auth query.Authorizer, shardIDs [
 			switch e.Op {
 			case influxql.EQ, influxql.NEQ, influxql.EQREGEX, influxql.NEQREGEX:
 				tag, ok := e.LHS.(*influxql.VarRef)
-				if !ok || tag.Val != "_name" {
+				if !ok || tag.Val != measurementKey {
 					return nil
 				}
 			}
 		}
 		return e
 	}), nil)
 
-	filterExpr := influxql.CloneExpr(cond)
-	filterExpr = influxql.Reduce(influxql.RewriteExpr(filterExpr, func(e influxql.Expr) influxql.Expr {
+	is, err := s.indexSetFromShards(ctx, shardIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	names, err := is.MeasurementNamesByExpr(nil, measurementExpr)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.tagValues(ctx, names, is, auth, cond)
+}
+
+// TagValuesByPredicate gets a slice of tag values, provided a flux predicate
+// that has been parsed into an influxql.Expr. It differs from TagValues in that
+// it works with flux predicates.
+func (s *Store) TagValuesByPredicate(ctx context.Context, auth query.Authorizer, shardIDs []uint64, cond influxql.Expr) ([]TagValues, error) {
+	is, err := s.indexSetFromShards(ctx, shardIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	measurementExpr := influxql.CloneExpr(cond)
+	measurementExpr = influxql.Reduce(influxql.RewriteExpr(measurementExpr, func(e influxql.Expr) influxql.Expr {
 		switch e := e.(type) {
 		case *influxql.BinaryExpr:
 			switch e.Op {
 			case influxql.EQ, influxql.NEQ, influxql.EQREGEX, influxql.NEQREGEX:
 				tag, ok := e.LHS.(*influxql.VarRef)
-				if !ok || influxql.IsSystemName(tag.Val) {
+				// Drop all system names except "_name" so that regex and negation
+				// operators work. The primary effect of this is dropping the "_tagKey"
+				// part of the expression for retrieving measurement names.
+				if !ok || (tag.Val != measurementKey && influxql.IsSystemName(tag.Val)) {
```

Comment:
I really hate this predicate collapsing - it's just wrong. e.g. all these evaluate the same:

`(_field == "foo") && (tag == "bar")`
`(_field != "foo") && (tag == "bar")`
`(_field == "foo") || (tag == "bar")`
`(_field != "foo") || (tag == "bar")`

See also https://github.com/influxdata/influxdb/pull/20793/files which fixed some of this for 1.x